codegen/templates.go

```diff
@@ -69,61 +69,65 @@ type templateHandler struct {
 }
 
 // Write takes one endpoint and uses a template to generate text
-// for it. This template is written to the given writer.
-func (h *templateHandler) Write(wr io.Writer, tmplType templateType, endpoint string, endpointInfo *framework.OASPathItem) error {
+// for it. This template is written to the given writer. It's exported
+// because it's the only method intended to be called by external callers.
+func (h *templateHandler) Write(wr io.Writer, endpoint string, endpointInfo *framework.OASPathItem, addedInfo *additionalInfo) error {
 	templatable, ok := h.templatableEndpoints[endpoint]
 	if !ok {
 		// Since each endpoint will have a code file and a doc file, let's cache
 		// the template-friendly version of the endpoint so it doesn't have to be
 		// converted into that format twice.
 		var err error
-		templatable, err = h.toTemplatable(endpoint, endpointInfo)
+		templatable, err = h.toTemplatable(endpoint, endpointInfo, addedInfo)
 		if err != nil {
 			return err
 		}
 		h.templatableEndpoints[endpoint] = templatable
 	}
-	return h.templates[tmplType].Execute(wr, templatable)
+	return h.templates[addedInfo.TemplateType].Execute(wr, templatable)
 }
 
-// toTemplatable converts the provided data into a struct that the template
-// can easily interpret.
-func (h *templateHandler) toTemplatable(endpoint string, endpointInfo *framework.OASPathItem) (*templatableEndpoint, error) {
-	parameters := collectParameters(endpointInfo)
+// toTemplatable does a bunch of work to format the given data into a
+// struct that has fields that will be idiomatic to use with Go's templating
+// language.
+func (h *templateHandler) toTemplatable(endpoint string, endpointInfo *framework.OASPathItem, addedInfo *additionalInfo) (*templatableEndpoint, error) {
+	parameters := collectParameters(endpointInfo, addedInfo)
 
 	// Sort the parameters by name so they won't shift every time
 	// new files are generated due to having originated in maps.
 	sort.Slice(parameters, func(i, j int) bool {
 		return parameters[i].Name < parameters[j].Name
 	})
 
-	// De-duplicate the parameters in place, because often parameters
-	// are at both the top-level and in the post body. This in-place
-	// approach is directly recommended here:
-	// https://github.com/golang/go/wiki/SliceTricks#in-place-deduplicate-comparable
-	j := 0
-	for i := 1; i < len(parameters); i++ {
-		if parameters[j] == parameters[i] {
-			continue
+	if len(parameters) > 2 {
+		// De-duplicate the parameters in place, because often parameters
+		// are at both the top-level and in the post body. This in-place
+		// approach is directly recommended here:
+		// https://github.com/golang/go/wiki/SliceTricks#in-place-deduplicate-comparable
+		j := 0
+		for i := 1; i < len(parameters); i++ {
```

Comment:
Can de-duplication be put into the `collectParameters` function? That seems like a common requirement for that function.

Reply:
Sure. When I do that, seems like I should move the sorting stuff too, and rename it to more of a "parseParameters" function. https://github.com/terraform-providers/terraform-provider-vault/pull/749/commits/76093acf660f8b36fa7456de6ffdc6d94cfcf960